consensus/istanbul/core/commit.go

```diff
@@ -96,6 +97,16 @@ func (c *core) handleCommit(msg *message, src istanbul.Validator) error {
 
 	c.acceptCommit(msg, src)
 
+	// Change to Prepared state if we've received enough PREPARE or COMMIT messages and we are in earlier state
+	// before Prepared state.
+	// Both of PREPARE and COMMIT messages are counted since hashlocked nodes in the previous round skip sending
+	// PREPARE messages.
+	if c.current.GetPrepareOrCommitSize() > 2*c.valSet.F() && c.state.Cmp(StatePrepared) < 0 {
```

Comment:
Do you have any reason why you skip the following condition `(c.current.IsHashLocked() && prepare.Digest == c.current.GetLockedHash())` ?

Reply:
If a node is hashlocked, it will do the same things when it received a `preprepare` message.

Reply:
I also wonder why it does not check the digest from the message, like it does in `handlePrepare`, like @jeongkyun-oh asked.

BTW, `LockHash` sounds a bit misleading since it just atomically updates the value of `roundState.lockedHash`. It just avoids concurrent read/write of the variable, not working as a usual lock.

Reply:
I think it is very rare for hashlocked node not to receive prepare msg.
So most hashlocked node can change its state to commit state by any prepare msg.
However, I think If there is a hashlocked node doesn't receive any prepare msg, this condition let it change the state to commit state by any commit msg.

Reply:
@jeongkyun-oh @ehnuje @KimKyungup As you mentioned, I added another condition for hash locked nodes. Thank you for letting me know.